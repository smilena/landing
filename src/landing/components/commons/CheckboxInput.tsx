import { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import styled from "styled-components";
import { setUserData, UserData } from '../../../redux-state/UserDataSlice';
import { RootState } from '../../../store';

export type CheckboxInputProps = {
  id: string;
  label: string;
  options?: Array<Option>;
}

type Option = {
  id: string;
  label: string;
}

const Components = {
  Wrapper: styled.div`
    display: flex;
    flex-direction: column;
  `,
  OptionWrapper: styled.div`
    display: flex;
    justify-content: space-between;
    margin: 4px;
  `,
  Label: styled.label`
    color: #504437;
    font-size: 18px;
    margin-bottom: 10px;
  `,
  Input: styled.input`
    accent-color: #504437;
  `,
}

export const CheckboxInput = ({id, label, options}: CheckboxInputProps) => {
  
  const dispatch = useDispatch();
  const userData: UserData = useSelector((state: RootState) => state.userData);
  const [checked, updateCheckedState] = useState(userData[id  as keyof UserData ] as Array<string> || []);

  useEffect(() => {
    dispatch(setUserData({id, value: checked}));
  }, [checked]);
  

  const onChangeHandler = (optionId: string) => {
    if (checked.includes(optionId)) {
      updateCheckedState(checked.filter((el: string) => el !== optionId));
    } else {
      updateCheckedState((currentState) => [...currentState, optionId])
    }
  }
  
  return (
    <Components.Wrapper>
      <Components.Label htmlFor={id}>{label}</Components.Label>
      {
       options && options.map((option, index) => 
          <Components.OptionWrapper key={index} >
            <Components.Label htmlFor={option.id}>{option.label}</Components.Label>
            <Components.Input
              type="checkbox"
              value={option.id}
              checked={checked.includes(option.id)}
              onChange={() => onChangeHandler(option.id)}
            />
          </Components.OptionWrapper>)
      }
    </Components.Wrapper>
  )
}
