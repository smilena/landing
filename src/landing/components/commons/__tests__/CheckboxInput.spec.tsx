import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import createMockStore from 'redux-mock-store';
import { CheckboxInput } from '../CheckboxInput';

const mockStore = createMockStore([]);

afterEach(cleanup);

describe('Given CheckboxInput component', () => {

  const mockProps = [
    {
      id: "bbq",
      label: "Zona BBQ"
    }
  ]
  
  test('should render correctly', () => {
    const state = {
      userData: {
        zones: ['bbq'],
      }
    };
    const store = mockStore(state);
    const { container } = render(
      <Provider store={store}>
        <CheckboxInput id={'floor'} label={'NÃºmero de piso'} options={mockProps}/>
      </Provider>);
    expect(container).toMatchSnapshot();
    expect(screen.getByRole('checkbox')).toBeInTheDocument();
    fireEvent.click(screen.getByRole('checkbox'));
  });
});
