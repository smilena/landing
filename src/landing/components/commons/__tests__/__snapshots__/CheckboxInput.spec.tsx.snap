// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given CheckboxInput component should render correctly 1`] = `
<div>
  <div
    class="sc-bczRLJ eaUlQc"
  >
    <label
      class="sc-dkzDqf ebBqBq"
      for="floor"
    >
      NÃºmero de piso
    </label>
    <div
      class="sc-gsnTZi kASbhL"
    >
      <label
        class="sc-dkzDqf ebBqBq"
        for="bbq"
      >
        Zona BBQ
      </label>
      <input
        class="sc-hKMtZM ijbipV"
        type="checkbox"
        value="bbq"
      />
    </div>
  </div>
</div>
`;
